crates/db_views/src/local_user_view.rs

```diff
@@ -13,132 +13,175 @@ use lemmy_db_schema::{
 };
 use lemmy_utils::error::{LemmyError, LemmyErrorExt, LemmyErrorType, LemmyResult};
 use std::future::{ready, Ready};
 
-enum ReadBy<'a> {
-  Id(LocalUserId),
-  Person(PersonId),
-  Name(&'a str),
-  NameOrEmail(&'a str),
-  Email(&'a str),
-  OAuthID(OAuthProviderId, &'a str),
+#[diesel::dsl::auto_type]
+fn joins() -> _ {
+  local_user::table
+    .inner_join(local_user_vote_display_mode::table)
+    .inner_join(person::table)
+    .inner_join(person_aggregates::table.on(person::id.eq(person_aggregates::person_id)))
 }
 
-enum ListMode {
-  AdminsWithEmails,
-}
-
-fn queries<'a>(
-) -> Queries<impl ReadFn<'a, LocalUserView, ReadBy<'a>>, impl ListFn<'a, LocalUserView, ListMode>> {
-  let selection = (
-    local_user::all_columns,
-    local_user_vote_display_mode::all_columns,
-    person::all_columns,
-    person_aggregates::all_columns,
-  );
-
-  let read = move |mut conn: DbConn<'a>, search: ReadBy<'a>| async move {
-    let mut query = local_user::table.into_boxed();
-    query = match search {
-      ReadBy::Id(local_user_id) => query.filter(local_user::id.eq(local_user_id)),
-      ReadBy::Email(from_email) => {
-        query.filter(lower(coalesce(local_user::email, "")).eq(from_email.to_lowercase()))
-      }
-      _ => query,
-    };
-    let mut query = query.inner_join(person::table);
-    query = match search {
-      ReadBy::Person(person_id) => query.filter(person::id.eq(person_id)),
-      ReadBy::Name(name) => query.filter(lower(person::name).eq(name.to_lowercase())),
-      ReadBy::NameOrEmail(name_or_email) => query.filter(
-        lower(person::name)
-          .eq(lower(name_or_email.to_lowercase()))
-          .or(lower(coalesce(local_user::email, "")).eq(name_or_email.to_lowercase())),
-      ),
-      _ => query,
-    };
-    let query = query
-      .inner_join(local_user_vote_display_mode::table)
-      .inner_join(person_aggregates::table.on(person::id.eq(person_aggregates::person_id)));
-
-    if let ReadBy::OAuthID(oauth_provider_id, oauth_user_id) = search {
-      query
-        .inner_join(oauth_account::table)
-        .filter(oauth_account::oauth_provider_id.eq(oauth_provider_id))
-        .filter(oauth_account::oauth_user_id.eq(oauth_user_id))
-        .select(selection)
-        .first(&mut conn)
-        .await
-    } else {
-      query.select(selection).first(&mut conn).await
-    }
-  };
-
-  let list = move |mut conn: DbConn<'a>, mode: ListMode| async move {
-    match mode {
-      ListMode::AdminsWithEmails => {
-        local_user::table
-          .inner_join(local_user_vote_display_mode::table)
-          .inner_join(person::table)
-          .inner_join(person_aggregates::table.on(person::id.eq(person_aggregates::person_id)))
-          .filter(local_user::email.is_not_null())
-          .filter(local_user::admin.eq(true))
-          .select(selection)
-          .load::<LocalUserView>(&mut conn)
-          .await
-      }
-    }
-  };
-
-  Queries::new(read, list)
-}
+type SelectionType = (
```

Comment:
The `SelectionType` is verbose unfortunately, but at least it can be auto-generated quickly.

Reply:
I think you should be able to use the use the [`Table::AllColumns`](https://docs.diesel.rs/2.2.x/diesel/prelude/trait.Table.html#associatedtype.AllColumns) type here to make this much more compact.

Reply:
Thx, I'll give that a try.

Reply:
That did it, thx!